File: kotlin/client/src/commonTest/kotlin/com/basecamp/actioncable/IdentifierTest.kt
package com.basecamp.actioncable

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class IdentifierTest {
    @Test
    fun `key`() {
        val identifiers =
            listOf(
                Identifier("RoomChannel") to """{"channel":"RoomChannel"}""",
                Identifier("RoomChannel", mapOf("id" to 42)) to """{"channel":"RoomChannel","id":42}""",
                Identifier("RoomChannel", mapOf("id" to 42, "since" to "yesterday")) to
                    """{"channel":"RoomChannel","id":42,"since":"yesterday"}""",
            )

        identifiers.forEach { (identifier, key) ->
            assertEquals(key, identifier.key)
            assertEquals(key, identifier.toString(), "toString should be the key")
        }
    }

    @Test
    fun `key refuses params it cannot encode`() {
        val identifier = Identifier("RoomChannel", mapOf("id" to Regex("nope")))

        assertFailsWith<ActionCableException> { identifier.key }
    }
}

File: typescript/tests/identifier.test.ts
import { describe, expect, it } from "vitest";
import { identifierKey, type Identifier } from "../src/index.js";

describe("identifierKey", () => {
  it("keys an identifier", () => {
    const identifiers: Array<[Identifier, string]> = [
      [{ channel: "RoomChannel" }, `{"channel":"RoomChannel"}`],
      [{ channel: "RoomChannel", params: { id: 42 } }, `{"channel":"RoomChannel","id":42}`],
      [
        { channel: "RoomChannel", params: { id: 42, since: "yesterday" } },
        `{"channel":"RoomChannel","id":42,"since":"yesterday"}`,
      ],
      [
        { channel: "RoomChannel", params: { since: "yesterday", id: 42 } },
        `{"channel":"RoomChannel","id":42,"since":"yesterday"}`,
      ],
    ];

    for (const [identifier, key] of identifiers) {
      expect(identifierKey(identifier), identifier.channel).toBe(key);
    }
  });

  it("refuses params it cannot encode", () => {
    expect(() => identifierKey({ channel: "RoomChannel", params: { id: 1n } })).toThrow(TypeError);
  });
});
